File: app/src/main/java/com/popalay/cardme/presentation/base/ViewModelPresenter.java
package com.popalay.cardme.presentation.base;

public abstract class ViewModelPresenter<T extends ViewModelView<M>,
        M extends BaseViewModel> extends BasePresenter<T> {

    protected final M viewModel;

    public ViewModelPresenter() {
        viewModel = createViewModel();
        getViewState().setViewModel(viewModel);
    }

    protected abstract M createViewModel();
}

File: app/src/main/java/com/popalay/cardme/presentation/base/ViewModelPresenter.kt
package com.popalay.cardme.presentation.base

abstract class ViewModelPresenter<T : ViewModelView<M>, M : BaseViewModel> : BasePresenter<T>() {

    protected val viewModel: M

    init {
        viewModel = createViewModel()
        viewState.setViewModel(viewModel)
    }

    protected  abstract fun createViewModel(): M
}
